opal/datatype/opal_datatype_copy.c

```diff
@@ -67,78 +55,104 @@ static size_t opal_datatype_memop_block_size = 128 * 1024;
         }                                                                                   \
     } while (0)
 
-/**
- * Overlapping memory regions
- */
+static inline void opal_datatype_accelerator_memcpy_impl(void *dest, const void *src, size_t size, int dev_id, 
+                                                          opal_accelerator_transfer_type_t type)
+{
+    int res = opal_accelerator.get_device(&dev_id);
+    if (OPAL_SUCCESS != res) {
+        /**
+         * Attempt the copy with 0 as dev_id. If we fail the copy then we'll abort anyway.
+         */
+        dev_id = 0;
+    }
+    res = opal_accelerator.memcpy(dev_id, dest, src, size, type);
+    if (OPAL_SUCCESS != res) {
+        opal_output(0, "Error in accelerator memcpy");
+        abort();
+    }     
+}
+
+static void *opal_datatype_accelerator_memcpy(void *dest, const void *src, size_t size)
+{
+    int res_src, res_dest, dev_id;
+    uint64_t flags;
+    
+    res_src = opal_accelerator.check_addr(src, &dev_id, &flags);
+    res_dest = opal_accelerator.check_addr(dest, &dev_id, &flags);
+    
+    if (res_src && res_dest) {
+        opal_datatype_accelerator_memcpy_impl(dest, src, size, dev_id, MCA_ACCELERATOR_TRANSFER_DTOD);
+    } else if (res_src) {
+        opal_datatype_accelerator_memcpy_impl(dest, src, size, dev_id, MCA_ACCELERATOR_TRANSFER_DTOH);
+    } else if (res_dest) {
+        opal_datatype_accelerator_memcpy_impl(dest, src, size, dev_id, MCA_ACCELERATOR_TRANSFER_HTOD);
+    } else {
+        MEMCPY(dest, src, size);
+    }
+    
+    return dest;
+}
+
+static inline void opal_datatype_accelerator_memmove_impl(void *dest, const void *src, size_t size, int dev_id, 
+                                                          opal_accelerator_transfer_type_t type)
+{
+    int res = opal_accelerator.get_device(&dev_id);
+    if (OPAL_SUCCESS != res) {
+        /**
+         * Attempt the move with 0 as dev_id. If we fail the move then we'll abort anyway.
+         */
+        dev_id = 0;
+    }
+    res = opal_accelerator.memmove(dev_id, dest, src, size, type);
+    if (OPAL_SUCCESS != res) {
+        opal_output(0, "Error in accelerator memmove");
+        abort();
+    }     
+}
+
+static void *opal_datatype_accelerator_memmove(void *dest, const void *src, size_t size)
```

Comment:
Why does this function make sense over just blindly calling memmove with type UNSPECIFIED?

Reply:
Because the NULL component doesn't do memcpy or memmove. Your ask was for check_addr to return 0 always, and I figured if we were going to do the check anyway, it would be more efficient to preserve that data and pass it as a type.

Reply:
Looking at all the uses, this function included, I may have been wrong.  It's weird to me that we're pushing complexity to the upper layers (read: many places) rather than handling it in the accelerator framework.  This is definitely one of the places that feels like we made the wrong decision about the complexity of the two implementation options.

Reply:
Changed the null component to do the memcpy and memmoves